### ISSUE TITLE
Add support for editions

### ISSUE THREAD

**Issue by ouillie** [tokio-rs/prost#1031]
[Editions](https://protobuf.dev/editions/overview/) are the future of protobuf syntax.

Currently, when you set `edition = "2023"` at the top of a `.proto` file, and run it through prost-build using `protoc_arg("--experimental_editions")`, it will panic on [this line](https://github.com/tokio-rs/prost/blob/04be6049ea676dde5dfac8a21da2559f4f380ed2/prost-build/src/code_generator.rs#L72) because the value of `FileDescriptorProto.syntax` [is `"editions"`](https://github.com/protocolbuffers/protobuf/blob/e5502c746ef62b3d76f589545803cfb46a5082ea/src/google/protobuf/descriptor.proto#L126). That's pretty much the extent of my investigation. I just figured I'd file this bug since support for editions doesn't seem to be tracked elsewhere.

This may sound like it's adding complexity to Prost, but it looks like it should actually simplify Prost once [Prototiller](https://protobuf.dev/editions/overview/#prototiller) is released, since that will effectively normalize all the different syntaxes and editions into a canonical representation using features, and Prost will only have to support that canonical implementation in the long term. It would be a somewhat heavy refactor up front, though.

**Comment by caspermeijn** [comment]
I agree, editions are important. However, due to backwards compatibility of `protoc` itself, we need to keep supporting `proto2` and `proto3` for some time. 

Are you willing to do the work to add this support?

**Comment by ouillie** [comment]
That makes sense. I won't have bandwidth for a couple months, but I'm down to contribute, especially after Prototiller is released.

**Comment by mzabaluev** [comment]
Up to spec support for editions requires at least the following: 

* Unknown fields, as discussed in #2.
* Open and closed enumeration behavior. I've started on this in #1079.

**Comment by esrauchg** [comment]
Drive by chiming in from someone on the Protobuf team at Google here.

Unfortunately open sourcing the prototiller tool is held up so we still don't have a date for that (sorry!), but in nearly all cases manually upgrading a file from Proto3 to Editions 2023 files shouldn't be too hard so there shouldn't be a very strong reason to wait

In general, the vision of Edition2023 is simply that it can reunify the syntaxes Proto2 and Proto3 and set the groundwork for future evolution in ~yearly Editions releases.  There's not really anything _extra_ in Edition 2023 that wasn't already there in Proto2 and Proto3. The ideal is that any Proto2 or Proto3 file should be able to move to Edition 2023 and have it be perfectly behavior preserving way (basically: start with Proto3, change it to Edition-2023, and the right options at the top of the file the the Prost codegen should be exactly the same as it was).

Ideally Editions 2023 support should only involve identifying any conditionals in the code generation which check exactly "is syntax == Proto2 or Proto3?" to decide things like "should this field be in an Option<>?" and decide in a more generalized way to also support Edition 2023 (which supports all proto2 and proto3 behavior under one syntax, and so a check on syntax can't decide those things anymore).

There's some argument against doing any extra conformance-to-spec work  for things like open-enums _specifically_ related to Prost advertising support of Edition2023 inputs, since ideally users should be able to move their Proto3 file to Edition2023 files and still get the same Prost gencode if they want. This does have oddities that if your implementation has preexisting non-conformant behavior on open-vs-closed enums, if you 'fix' that behavior only on Edition2023 it will make it a breaking change for users to upgrade their .proto files, so the strategy we recommend is more "make Edition2023 behavior whatever the corresponding preexisting proto2/proto3 behavior was, then _separately_ consider fixing nonconformant behavior of things like open enums on both Proto3 and Edition2023 together a broader "conformance to intended spec" topic.

I glanced at the Prost generator code, I only immediately see 2 conditionals against .syntax in code_generator.rs that would need to be fixed for it to handle Edition2023 inputs:
https://github.com/tokio-rs/prost/blob/master/prost-build/src/code_generator.rs#L1037

The main work that would need to happen for Prost to support Edition 2023 is for those 2 conditionals to honor Feature Resolution: this relates to the behavior where under Editions the idea is that you can set if fields should be treated as optional or not by default at any of the file > message > nested message > individual field level scopes and it applies downward within the scope. For every feature, we also publish values of them that match the Proto2/Proto3 behavior, so the intended design is that you can have your code generator operate _only_ on the feature notion and you'll get support for all possible syntaxes.

Protoc plugins written in C++ get this feature-scoping/cascading behavior for free since we implemented it in the descriptor.h that we publish, but code generators written in other languages that have done their own bootstrap on descriptor.proto (like Prost has) directly need to implement that. It generally amounts to a DFS to propagate the feature values downward in scopes, but it could e.g. a decision on a field could on demand traverse upwards until you find the narrowest scope that decides what the optionality of that field should be. Once you have that "feature cascading" behavior in place, I believe Prost supporting Edition 2023 is only like a ~3 line delta.

There's more info about Feature Resolution here:
https://protobuf.dev/editions/implementation/#:~:text=specific%20test%20cases-,Feature%20Resolution,-Editions%20use%20lexical&text=specific%20test%20cases-,Feature%20Resolution,-Editions%20use%20lexical

**Comment by mzabaluev** [comment]
Field presence is going to be challenging. Starting from edition 2023, all fields have explicit presence (equivalent to declaring the fields optional in proto3) by default. There is pushback even about message-typed fields being represented with an `Option`, as seen in #763 and a number of duplicate issues.

I see two ways to deal with this:

1. Maintain the current approach: keep all struct fields public, represent explicitly present fields with `Option`, alleviate the pain with getter and setter methods for scalar fields.
2. Follow what Google did with the Go bindings and change the code generator to produce opaque structs, with private fields and public accessor methods. In-memory representation for field presence can be optimized with presence flags.

Advantages:

1. A simple and straightforward representation. The rich and standard `Option` API can be used to handle the potentially missing value in a way most appropriate to the application.
2. Memory layout of message structs can be more compact. The generated structs may be more suitable as application data domain types (however, this is not generally achievable due to validity constraints out of scope for Protobuf).

Disadvantages:

1. As explicit presence becomes the norm, the convenience of destructuring diminishes: you have to match for the `None` case on each field of interest.
2. It's impossible to destructure or use struct initializer syntax at all. The latter is not Protobuf-compliant, though; forward-compatible code should use [builders](https://github.com/tokio-rs/prost/issues/399).

**Comment by esrauchg** [comment]
Ah; I think officially Protobuf is just not really optimal fit for wasm/FFI bridge style usecases; Protobuf is focused on ensuring behavior in the face of skewed schemas between sender/receiver so that you can safely change schema over time in distributed system; by contrast FFI you can know for sure there's no schema skew. Generally any dedicated FFI technologies will suit you better for that, as well as some other serialization technologies like capnproto or flatbuffers.

This is getting a bit further the weeds about a different topic that this Prost Issue; if there's any further questions that the Protobuf team can help with (both from Prost contributors or any other users), please feel free open an issue on the Protobuf github and we are happy to discuss further there, or

**Comment by mzabaluev** [comment]
> > NB: ..Default::default() update syntax does not work with non-exhaustive structs.
> 
> Maybe I'm misunderstanding but what I was thinking of does work, [here's an example in the Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=01d0ca708b8d9a991066eeda6b34c04b)

That only works within the same crate, not with exported APIs.

> ```
> const CHILD_DEFAULT: Child = Default::default();
> impl Parent {
>     fn c(&self) -> &Child {
>         self.c.as_ref().unwrap_or(&CHILD_DEFAULT)
>     }
> }
> ```

Not as simple as this: `Default::default` is not const, and you can't generate a const initializer in all cases without getting into weeds with `LazyLock` or similar, because e.g. `HashMap` cannot be const-initialized without sacrifices in security.

While this seems feasible, I'm not sure it's an architectural choice that Prost should adopt, being a minimal-cost Protobuf marshalling/unmarshalling implementation. Maybe with open structs, the getters could be considered amenities with an acceptable potential for synchronization blockage? The developers would need to be educated about this. I'm assuming code generated for `no_std` can only use const-initializable types, so at least that is not a concern.

**Comment by ensc** [comment]
Would it be possible to emulate `optional` with features/metadata?  E.g.

```protobuf
message Foo {
    int32 id = 1 [(prost.optional) = false];
    string name = 1 [(prost.optional) = true];
}
```

which would generate

```rust
struct Foo {
    id: i32,
    name: Option<String>,
}
```
